Group members:

Juan Sevilla Fernández

Álvaro Cabrerizo de Garnica

In [1]:
import pandas as pd
import numpy as np

In [2]:
# You can download the data from
# https://github.com/joashxu/prosper-loan-data/blob/master/dataset/prosperLoanData.csv

# Dataset description
# https://github.com/joashxu/prosper-loan-data/tree/master/dataset

# Read the Prosper Loan Data
df = pd.read_csv("/Users/juan/Documents/Juan's MacBook Pro/CUNEF/Quinto/Credit Risk Analytics/data/prosperLoanData.csv")

In [3]:
df.columns

Index(['ListingKey', 'ListingNumber', 'ListingCreationDate', 'CreditGrade',
       'Term', 'LoanStatus', 'ClosedDate', 'BorrowerAPR', 'BorrowerRate',
       'LenderYield', 'EstimatedEffectiveYield', 'EstimatedLoss',
       'EstimatedReturn', 'ProsperRating (numeric)', 'ProsperRating (Alpha)',
       'ProsperScore', 'ListingCategory (numeric)', 'BorrowerState',
       'Occupation', 'EmploymentStatus', 'EmploymentStatusDuration',
       'IsBorrowerHomeowner', 'CurrentlyInGroup', 'GroupKey',
       'DateCreditPulled', 'CreditScoreRangeLower', 'CreditScoreRangeUpper',
       'FirstRecordedCreditLine', 'CurrentCreditLines', 'OpenCreditLines',
       'TotalCreditLinespast7years', 'OpenRevolvingAccounts',
       'OpenRevolvingMonthlyPayment', 'InquiriesLast6Months', 'TotalInquiries',
       'CurrentDelinquencies', 'AmountDelinquent', 'DelinquenciesLast7Years',
       'PublicRecordsLast10Years', 'PublicRecordsLast12Months',
       'RevolvingCreditBalance', 'BankcardUtilization',
       'Availa

In [4]:
df.shape

(113937, 81)

In [5]:
df.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380  ...         -133.18                0.0   
1        0.0920       0.0820  ...            0.00                0.0   
2        0.2750       0.2400  ...          -24.20                0.0   
3        0.0974       0.0874  ...         -108.01                0.0   
4        0.2085       0.1985  ...          -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

## 1

Display the "LoanStatus" and "Investors" columns for the row with "ListingNumber" = 816108

In [6]:
## WRITE YOUR CODE HERE ##
df1 = df.loc[df['ListingNumber'] == 816108]
df1[['ListingNumber',"LoanStatus", "Investors"]]

ListingNumber LoanStatus  Investors
44         816108    Current          1

## 2

Create a new column called "SingleInvestor" which takes the value True if the column "Investors" is 1, and takes the value False otherwise

In [7]:
## WRITE YOUR CODE HERE ##
df['SingleInvestor'] = (df['Investors']== 1) & (df['Investors']>=1)
df[['Investors',"SingleInvestor"]].head(10)

Investors  SingleInvestor
0        258           False
1          1            True
2         41           False
3        158           False
4         20           False
5          1            True
6          1            True
7          1            True
8          1            True
9          1            True

## 3

Count how many rows have NA as ProsperScore, and count how many rows have a ProsperScore higher than 8

In [8]:
## WRITE YOUR CODE HERE ##
df3NAs = df["ProsperScore"].isna().sum()
print( "There are " + str(df3NAs)+ " instances where ProsperScore is NA")

There are 29084 instances where ProsperScore is NA


In [9]:
df3 = df[df["ProsperScore"] > 8]
print( "There are " + str(len(df3))+ " instances where ProsperScore is higher than 8")

There are 13117 instances where ProsperScore is higher than 8


## 4

Display how many rows and mean "Investors" we have per CreditGrade

In [10]:
## WRITE YOUR CODE HERE ##
df.groupby(["CreditGrade"])[[ "Investors"]].aggregate(['count', 'mean'])


Investors            
                count        mean
CreditGrade                      
A                3315  186.689291
AA               3509  188.102023
B                4389  168.241285
C                5649  111.960701
D                5153   84.656705
E                3289   48.395257
HR               3508   30.510547
NC                141   28.617021

## 5

The "LoanFirstDefaultedCycleNumber" column captures the month in which the loan first defaulted. Create a column called fraud, that will be True if the loan was defaulted in less than 4 months. Calculate what is the fraud rate in the portfolio based on that new column.

In [11]:
df['Fraud'] = False
df.loc[df['LoanFirstDefaultedCycleNumber'] < 4, 'Fraud'] = True
Fr_rate = df['Fraud'].mean()
print("The fraud rate in the porfolio is " + str(round(Fr_rate*100, 2))+ "%")

The fraud rate in the porfolio is 0.49%


## 6

Create a new column called is_student, which will be true if the column "Occupation" contains the text "Student". Count how many rows and what is the fraud rate based on the is_student column.

In [12]:
## WRITE YOUR CODE HERE ##
df["is_student"] = df['Occupation'].str.contains('Student', case=False, na=False)
df6 = df[['Occupation',"is_student", "Fraud"]].loc[df['is_student'] == True]

Fr_st = df6['Fraud'].mean()

print("There are " + str(len(df6)) + " rows where the loan is given to a Student")
print("The fraud rate for students in the protfolio is " + str(round(Fr_st*100, 2))+ "%")

There are 699 rows where the loan is given to a Student
The fraud rate for students in the protfolio is 0.43%


## 7

Study the fraud rate versus the "IncomeRange" columns. Is there any insight we can learn from it?

In [13]:
## WRITE YOUR CODE HERE ##
round(df.groupby(["IncomeRange"])[["Fraud"]].mean()*100, 2).sort_values(by = "Fraud", ascending = False)

Fraud
IncomeRange          
Not displayed    1.46
$0               1.13
Not employed     0.62
$1-24,999        0.56
$50,000-74,999   0.46
$25,000-49,999   0.45
$75,000-99,999   0.41
$100,000+        0.21

#### EXPLAIN YOUR INSIGHTS HERE 
Clearly there is a higher Fraud rate for lower or no incomes. <br> 
It is understanable that people with lower income will look to borrow money without the intention of ever paying it back.  <br> 
The only Income Range that does not follow the trend is the $50,000-74,999	bucket as it has a slightly higher rate than the 25,000-49,999 bucket

## 8

Compute the correlation between the features in the next cell. Is there any insight we can learn from the correlation?

In [14]:
feats = ["EmploymentStatusDuration",
         "OpenRevolvingAccounts", 
         "OpenCreditLines", 
         "CreditScoreRangeLower", 
         "CreditScoreRangeUpper"]

In [15]:
## WRITE YOUR CODE HERE ##
df[feats].corr()

EmploymentStatusDuration  OpenRevolvingAccounts  \
EmploymentStatusDuration                  1.000000               0.154702   
OpenRevolvingAccounts                     0.154702               1.000000   
OpenCreditLines                           0.139730               0.885448   
CreditScoreRangeLower                     0.081134               0.210659   
CreditScoreRangeUpper                     0.081134               0.210659   

                          OpenCreditLines  CreditScoreRangeLower  \
EmploymentStatusDuration         0.139730               0.081134   
OpenRevolvingAccounts            0.885448               0.210659   
OpenCreditLines                  1.000000               0.148210   
CreditScoreRangeLower            0.148210               1.000000   
CreditScoreRangeUpper            0.148210               1.000000   

                          CreditScoreRangeUpper  
EmploymentStatusDuration               0.081134  
OpenRevolvingAccounts                  0.210659  
OpenCreditLines                        0.148210  
CreditScoreRangeLower                  1.000000  
CreditScoreRangeUpper                  1.000000

#### EXPLAIN YOUR INSIGHTS HERE 

There are only two correlations that exceed 80% in absolute value. These are: <ol>
        <li> OpenCreditLines and OpenRevolvingAccounts, which means that the more credit lines the the account has, the more likely and more often it is for them to miss payment on these creidit lines or cards (this is known as revolving)</li>
       <li>  There is a perfect correlation between CreditScoreRangeLower and CreditScoreRangeUpper, which means the the range for the credit score has a fixed width and these two limits go up or down in perfect parity and always by the same amount.</li></ol><br>
      
      
The next strongest correlation is 21% which is between the nuber of OpenRevolvingAccounts and the CreditScoreRange. This is surpising as it suggests that, although to a lower degree,the higher the credit score, the higher the number of revolving lines (late payments on credit lines). It is surpising because a "bad" client factor should not positvely correlate with the credit score.
        
        